### ISSUE TITLE
Tooltip HTML

### ISSUE THREAD

**Issue by aniznik1** [amcharts/amcharts5#35]
Is it possible to add custom HTML to a tooltip?

Similarly to `series.tooltipHTML` in amcharts 4

**Comment by derekbking** [comment]
I was wondering the same thing. Considering amcharts5 is an overhaul moving towards canvas rendering as opposed to built in DOM elements, it feels like this feature wouldn't be as natural to support.

I will be sad to see this feature go if there are no plans to support it. Creating nice looking tooltips was much easier for me using markup.

I have been looking into getting this feature to work on one of my charts by manually displaying and updating markup based on events emitted from amcharts. I haven't quite gotten it to work right, but I think it is a feasible workaround. Albeit messy.

**Comment by sassomedia** [comment]
I think even being able to add a Container into the Tooltip would help in the meantime. That could allow for some basic layouting of elements.

**Comment by martynasma** [comment]
Well, `Tooltip` is a `Container`. You could technically add any stuff into it.

**Comment by sassomedia** [comment]
Thanks, that seems to work for me:
```javascript
const ttip = am5.Tooltip.new(root, {});
ttip.get("background").set("fill", am5.color(0xff0000));

const container = am5.Container.new(root, { layout: root.verticalLayout });
const label1 = am5.Label.new(root, { text: "Label text" });
const label2 = am5.Label.new(root, { text: "More label text" });

container.children.push(label1);
container.children.push(label2);
ttip.children.push(container);
```
With the `root.gridLayout` that already enables some possibilities...

**Comment by rzseattle** [comment]
Looks gr8. 

Last question is how to bind data to labels?

**Comment by sassomedia** [comment]
I don't know if that's directly possible but for an am4-chart with a fairly custom tooltip I used something like the following for a date-based chart. Making use of the cursor position and look up data based on that:
```javascript
const dateAxis = chart.xAxes.getIndex(0);

function showSeriesTooltip() {
  const { tooltipDate } = dateAxis;

  if (tooltipDate) {
    const series = chart.series.getIndex(0);
    // find data based on tooltipDate
    const dataSet = series.data.find(item => item.date === tooltipDate);

    // assemble custom tooltip content in here, e.g.
    series.tooltipHTML = `<table>
      <tr>
        <td style="padding-right:10px;">${dataSet.id}</td>
        <td style="text-align:right;">${dataSet.name}</td>
      </tr>
    <table>`;
  }
}

chart.cursor.events.on("cursorpositionchanged", showSeriesTooltip);

/* ---------- or, if it's for a chart tooltip ----------- */

function getOrientation() {
  const axisTooltipX = dateAxis.tooltip.properties.x;
  const chartMidPoint = chart.pixelWidth / 2;
  const orientation = axisTooltipX > chartMidPoint ? "right" : "left";

  return { axisTooltipX, orientation };
}

function showChartTooltip() {
  const { tooltipDate } = dateAxis;

  if (tooltipDate) {
    // find data based on tooltipDate
    const dataSet = chart.data.find(item => item.date === tooltipDate);
    const { axisTooltipX, orientation } = getOrientation();

    // assemble custom tooltip content in here, e.g.
    chart.tooltipHTML = `<table>
      <tr>
        <td style="padding-right:10px;">${dataSet.id}</td>
        <td style="text-align:right;">${dataSet.name}</td>
      </tr>
    <table>`;

    chart.tooltip.animate([{ property: "x", to: axisTooltipX }], 100);
    chart.tooltip.pointerOrientation = orientation;
  }
}

chart.cursor.events.on("cursorpositionchanged", showChartTooltip);
```
I have not tried this approach yet with am5 but I would think it could probably work similarly. Of course, substituting the `tooltipHTML` with the container approach mentioned above. If it's an XY chart, I think there should always be a way to get the position to map that to the corresponding data.

**Comment by rzseattle** [comment]
Solution based on prev container proposition: 

```js    
    const ttip = am5.Tooltip.new(root, {});

    let currTooltipYear = null;
    ttip.events.on("dataitemchanged", (n) => {
        const data: any = ttip.dataItem.dataContext;
        if (data.year !== currTooltipYear) {
            ttip.children.clear();

            const container = am5.Container.new(root, {
                layout: am5.GridLayout.new(root, {
                    maxColumns: 2,
                }),
                width: 200,
            });

           // .... add chrildren to container with current data


            ttip.children.push(container);
        }
    });

````

**Comment by JohnnyKmLi** [comment]
Why does my tooltip not show when I hover over the bullet?
`   const root = am5.Root.new('chart');
    root.dateFormatter.setAll({
      dateFormat: 'yyyy-MM-dd',
      dateFields: ['valueX']
    });

    const chart = root.container.children.push(
      XYChart.new(root, {
        panX: false,
        panY: false,
        wheelY: 'zoomXY',
        layout: root.verticalLayout
      })
    );
    chart.get('colors').set('step', 2);

    const xAxis = chart.xAxes.push(this.createXAxisObject(root, chart, AxisType.Date));
    const yAxis = chart.yAxes.push(this.createYAxisObject(root, chart, AxisType.Category));

    this.createSeries(chart, root, xAxis, yAxis);

    return chart;
  }

  createXAxisObject(root: am5.Root, chart: XYChart, type: AxisType): DateAxis<AxisRenderer> {
    let axisObject: DateAxis<AxisRenderer>;
    if (type === AxisType.Date) {
      const startDateRange = new Date();
      const endDateRange = new Date();
      endDateRange.setMonth(endDateRange.getMonth() + 6);

      const xRenderer = AxisRendererX.new(root, {
        minGridDistance: 50,
        strokeOpacity: 1,
        strokeWidth: 1
      });
      xRenderer.grid.template.setAll({ strokeDasharray: [5, 5] });
      axisObject = DateAxis.new(root, {
        min: startDateRange.getTime(),
        max: endDateRange.getTime(),
        baseInterval: { timeUnit: 'month', count: 1 },
        renderer: xRenderer
      });
      // xRenderer.grid.template.setAll({ location: 0.5 });
    }

    return axisObject;
  }

  createYAxisObject(root: am5.Root, chart: XYChart, type: AxisType): CategoryAxis<AxisRenderer> {
    let axisObject: CategoryAxis<AxisRenderer>;
    if (type === AxisType.Category) {
      const yRenderer = AxisRendererY.new(root, {
        minGridDistance: 20,
        strokeOpacity: 1,
        strokeWidth: 1
      });
      yRenderer.grid.template.setAll({ location: 0.5, strokeDasharray: [5, 5] });
      axisObject = CategoryAxis.new(root, {
        maxDeviation: 0,
        renderer: yRenderer,
        categoryField: 'criticality'
      });
    }

    return axisObject;
  }

  createSeries(
    chart: XYChart,
    root: am5.Root,
    xAxis: DateAxis<AxisRenderer>,
    yAxis: CategoryAxis<AxisRenderer>
  ): void {
    const ttip = am5.Tooltip.new(root, {});
    ttip.get('background').set('fill', am5.color('#ff0000'));
    const container = am5.Container.new(root, { layout: root.verticalLayout });
    const label1 = am5.Label.new(root, { text: 'Label text' });
    const label2 = am5.Label.new(root, { text: 'More label text' });

    container.children.push(label1);
    container.children.push(label2);
    ttip.children.push(container);

    const series = chart.series.push(
      LineSeries.new(root, {
        name: 'lineSeries',
        calculateAggregates: true,
        stroke: am5.color(0xffffff),
        xAxis,
        yAxis,
        valueXField: 'date',
        categoryYField: 'criticality',
        valueField: 'value',
        tooltip: Tooltip.new(root, {})
      })
    );
    series.strokes.template.setAll({ forceHidden: true });

    const circleTemplate: Template<any> = am5.Template.new({ radius: 5 });

    series.bullets.push(() => {
      const circle = am5.Circle.new(
        root,
        {
          stroke: series.get('stroke'),
          strokeWidth: 3,
          fill: series.get('fill'),
          tooltipY: -10,
          tooltip: ttip
        },
        circleTemplate
      );
      return am5.Bullet.new(root, { sprite: circle });
    });

    series.get('tooltip').label.set('text', 'here I am!');
    series.set('heatRules', [
      {
        target: circleTemplate,
        min: 5,
        max: 35,
        dataField: 'value',
        key: 'radius'
      }
    ]);

    series.data.processor = am5.DataProcessor.new(root, { dateFields: ['date'], dateFormat: 'yyyy-MM-dd' });

    // same date larger value first otherwise render bullet will overlap
    const data = [
      {
        date: '2021-12-02',
        criticality: 'High',
        value: 199
      },
      {
        date: '2022-02-11',
        criticality: 'Medium',
        value: 1990
      },
      {
        date: '2022-03-14',
        criticality: 'Low',
        value: 2990
      },
      {
        date: '2022-03-14',
        criticality: 'Low',
        value: 990
      }
    ];

    series.data.setAll(data);

    yAxis.data.setAll([
      { criticality: 'Not Applicable' },
      { criticality: 'Very Low' },
      { criticality: 'Low' },
      { criticality: 'Medium' },
      { criticality: 'High' },
      { criticality: 'Very High' }
    ]);

    chart.appear(1000, 100);
  }
}`

**Comment by martynasma** [comment]
@JohnnyKmLi Let's keep this feature request clean. Please post separate issue.

**Comment by chis0m** [comment]
@martynasma  **tooltipHTML** is a important feature to have. We currently have a usecase to implment it but sad to find out it is not in amChart5

**Comment by martynasma** [comment]
Implemented in [5.2.11](https://github.com/amcharts/amcharts5/releases/tag/5.2.11).

## [5.2.11] - 2022-07-18

### Added
- A new setting `deactivateRoot` (default: `true`) added to `Modal`. Indicates if any interactivity on the underlying chart should be disabled when modal is open.
- A new setting `html` added to `Container`. Set it to a string with HTML to be used as container's content. Will work on any element inheriting from `Container`, e.g. `Label`. [More info](https://www.amcharts.com/docs/v5/concepts/common-elements/html-content/).
- A new setting `labelHTML` added to `Tooltip`. If set, will ne used as HTML content for the tooltip. [More info](https://www.amcharts.com/docs/v5/concepts/common-elements/html-content/).
- A new setting `tooltipHTML` added to `Sprite`. If set, will ne used as HTML content for the tooltip when element is hovered. [More info](https://www.amcharts.com/docs/v5/concepts/common-elements/html-content/).
- `wheelZoomPositionX` and `wheelZoomPositionY` added to `XYChart`. This value is not set by default, but you can use it to fix wheel zooming, to the end (if you set it to 1) or to the start (0) or middle (0.5). In `StockChart`, it is set to 1 by default, as this is common practice in financial charts.
- "Volume-Weighted Average Price (VWAP)" indicator added to `StockChart`.

### Changed
- Setting `id` setting when creating an object will not register it in `am5.registry.entitiesById` immediatelly, without waiting for the next frame.

### Fixed
- `RadialGradient`'s `x` and `y` settings were being ignored.
- Labels as bullets with `oversizedBehavior` set were not being displayed properly in some cases.
- `DateAxis` with `startLocation = 0.5` and `endLocation = 0.5` and single value only was showing a really big date range.
- Resizing `PieChart` to a very small size with a legend positioned on the left or right could cause some anomaly in layout.
- Updating `CategoryAxis` data with less items than there was originally was causing old columns/bullets to be shown.
- Setting `verticalScrollbar` on `Container` to `undefined` to remove a previously set scrollbar was not working.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
